satpy/readers/modis_l3.py

```diff
@@ -0,0 +1,132 @@
+------------
+
+The ``modis_l3`` reader reads Modis L3 products in hdf-eos format.
+
+There are multiple level 3 products, including some on sinusoidal grids and some on the climate modeling grid (CMG).
+This reader supports the CMG products at present, and the sinusoidal products will be added if there is demand.
+
+The reader has been tested with:
+    - MCD43c*: BRDF/Albedo data, such as parameters, albedo and nbar
+    - MOD09CMG: Surface Reflectance on climate monitoring grid.
+
+To get a list of the available datasets for a given file refer to the "Load data" section in :doc:`../reading`.
+
+"""
+import logging
+
+from pyresample import geometry
+
+from satpy.readers.hdfeos_base import HDFEOSGeoReader
+
+logger = logging.getLogger(__name__)
+
+
+class ModisL3GriddedHDFFileHandler(HDFEOSGeoReader):
+    """File handler for MODIS HDF-EOS Level 3 CMG gridded files."""
+
+    def _get_res(self):
+        """Compute the resolution from the file metadata."""
+        gridname = self.metadata["GridStructure"]["GRID_1"]["GridName"]
+        if "CMG" not in gridname:
+            raise ValueError("Only CMG grids are supported")
+
+        # Get the grid resolution from the grid name
+        pos = gridname.rfind("_") + 1
+        pos2 = gridname.rfind("Deg")
+
+        # Some products don't have resolution listed.
+        if pos < 0 or pos2 < 0:
+            self.resolution = 360. / self.ncols
+        else:
+            self.resolution = float(gridname[pos:pos2])
+
+    def _sort_grid(self):
+        """Get the grid properties."""
+        # First, get the grid resolution
+        self._get_res()
+
+        # Now compute the data extent
+        upperleft = self.metadata["GridStructure"]["GRID_1"]["UpperLeftPointMtrs"]
+        lowerright = self.metadata["GridStructure"]["GRID_1"]["LowerRightMtrs"]
+
+        # For some reason, a few of the CMG products multiply their
+        # decimal degree extents by one million. This fixes it.
+        if lowerright[0] > 1e6:
+            upperleft = tuple(val / 1e6 for val in upperleft)
+            lowerright = tuple(val / 1e6 for val in lowerright)
+
+        return (upperleft[0], lowerright[1], lowerright[0], upperleft[1])
+
+
+    def __init__(self, filename, filename_info, filetype_info, **kwargs):
+        """Init the file handler."""
+        super().__init__(filename, filename_info, filetype_info, **kwargs)
+
+        # Initialise number of rows and columns
+        self.nrows = self.metadata["GridStructure"]["GRID_1"]["YDim"]
+        self.ncols = self.metadata["GridStructure"]["GRID_1"]["XDim"]
+
+        # Get the grid name and other projection info
+        self.area_extent = self._sort_grid()
+
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+        logger.debug("Available_datasets begin...")
+
+        ds_dict = self.sd.datasets()
+
+        yield from super().available_datasets(configured_datasets)
```

Comment:
I don't think you can blindly yield from this unless you make two assumptions:

1. There will never be more than one file passed to the reader at a time.
2. There will never be a dataset configured in the YAML.

For 1, you are likely redefining every dataset that is in every file since this `yield from` will produce all the datasets from other file handlers. If you have two files passed to the reader (meaning two file handlers are created) and they had the same variables then you'd get the first set from the first file handler which would then be yielded from this `yield from` line, then the same variables are yielded again in the below for loop. If two files with the same variables are never read at the same time then maybe this isn't a problem, but it seems like a future "gotcha".

For 2, if there is a chance that the variable (file_key) defined in the YAML isn't in the file being read then instead of a `yield` from, we should maybe be checking that the filetype matches this current file handler and that the variable exists in `ds_dict`. Also, if it does exist in `ds_dict` (it is part of the file) then we may not want to yield a copy of it in the below for loop.

I typically keep a `set()` object around of variable names that I've seen already and don't yield them again.

Reply:
I've tried to address this in a new commit now - hopefully it satisfies what you're requesting.
Coverage is poor right now as `configured_datasets` is empty. Will try to fix 'em.

Reply:
I'm still reading the rest of this, but you'll want to remove this `yield from` line as you're handling `configured_datasets` below.

Reply:
Hah yes, I was just debugging that now.